ga4gh/gff3Parser.py

```diff
@@ -0,0 +1,321 @@
+        self.score = score
+        self.strand = strand
+        self.frame = frame
+        self.attributes = copy.deepcopy(attributes)
+        self.gff3Set = gff3Set
+        self.lineNumber = lineNumber
+        self.parents = set()
+        self.children = set()
+
+    @staticmethod
+    def __dotIfNone(val):
+        return val if val is not None else '.'
+
+    def __attributeStr(self, name):
+        """
+        Return name=value for a single attribute
+        """
+        return "{}={}".format(
+            _encodeAttr(name),
+            ",".join([_encodeAttr(v) for v in self.attributes[name]]))
+
+    def __attributeStrs(self):
+        """
+        Return name=value, semi-colon-separated string for attributes,
+        including url-style quoting
+        """
+        return ";".join([self.__attributeStr(name)
+                         for name in self.attributes.iterkeys()])
+
+    def __str__(self):
+        """
+        Return the object as a valid GFF3 record line.
+        """
+        return "\t".join([self.seqname, self.source, self.type,
+                          str(self.start), str(self.end),
+                          self.__dotIfNone(self.score),
+                          self.__dotIfNone(self.strand),
+                          self.__dotIfNone(self.frame),
+                          self.__attributeStrs()])
+
+    @property
+    def featureId(self):
+        """
+        ID attribute or None if records doesn't have an ID
+        """
+        featId = self.attributes.get("ID")
+        if featId is not None:
+            featId = featId[0]
+        return featId
+
+    def getRequiredAttribute(self, name):
+        "get a require attributes list of values, error if not found"
+        values = self.attributes.get(name)
+        if values is None:
+            raise GFF3Exception(
+                "required attribute not found: {}".format(name),
+                self.gff3Set.fileName, self.lineNumber)
+        return values
+
+
+class Gff3Set(object):
+    """
+    A set of GFF3 sequence annotations
+    """
+    def __init__(self, fileName=None):
+        self.fileName = fileName
+        self.roots = set()     # root nodes (those with out parents)
+        # index of features by id. GFF3 allows disjoint features with
+        # the same id.  None is used to store features without ids
+        self.byFeatureId = collections.defaultdict(list)
+
+    def add(self, feature):
+        """
+        Add a feature record
+        """
+        # None featureId allowed
+        self.byFeatureId[feature.featureId].append(feature)
+
+    def __linkFeature(self, feature):
+        """
+        Link a feature with it's parents.
+        """
+        parentIds = feature.attributes.get("Parent")
+        if parentIds is None:
+            self.roots.add(feature)
+        else:
+            for parentId in parentIds:
+                self.__linkToParent(feature, parentId)
+
+    def __linkToParent(self, feature, parentId):
```

Comment:
gff3Parser.py is import from my code base.  I religiously only
make things `protected' (single underscore) if they are designed
to be overridden.  Overriding methods that are not intended to
be overridden leaves to implicit dependencies that are easy to
break.  These are sometimes behavioral dependencies that having
compiler will not help with.

Inheritance polymorphism is one of the pitfalls of OOP.


Jerome Kelleher <notifications@github.com> writes:
> In ga4gh/gff3Parser.py:
> 
> > +        """
> > +        # None featureId allowed
> > +        self.byFeatureId[feature.featureId].append(feature)
> > +
> > +    def __linkFeature(self, feature):
> > +        """
> > +        Link a feature with it's parents.
> > +        """
> > +        parentIds = feature.attributes.get("Parent")
> > +        if parentIds is None:
> > +            self.roots.add(feature)
> > +        else:
> > +            for parentId in parentIds:
> > +                self.__linkToParent(feature, parentId)
> > +
> > +    def __linkToParent(self, feature, parentId):
> 
> The only issue I have with private methods and attributes is that things can
> get awkward when you start subclassing. This is why I tend not to bother with
> them these days. It's up to you though, it's not a strong opinion.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub*
>